static/app/components/core/compactSelect/compactSelect.spec.tsx

```diff
@@ -1,10 +1,104 @@
-import {Fragment} from 'react';
+import {Fragment, useState} from 'react';
+import {expectTypeOf} from 'expect-type';
 
 import {render, screen, userEvent, waitFor} from 'sentry-test/reactTestingLibrary';
 
-import {CompactSelect} from './';
+import {CompactSelect, type SelectOption} from './';
 
 describe('CompactSelect', () => {
+  describe('types', () => {
+    it('should enforce correct types for onChange for SingleSelect', () => {
+      void (
+        <CompactSelect
+          value="opt_one"
+          onChange={option => {
+            expectTypeOf(option).toEqualTypeOf<SelectOption<'opt_one' | 'opt_two'>>();
+          }}
+          closeOnSelect={option => {
+            expectTypeOf(option).toEqualTypeOf<SelectOption<'opt_one' | 'opt_two'>>();
+            return true;
+          }}
+          options={[
+            {value: 'opt_one', label: 'Option One'},
+            {value: 'opt_two', label: 'Option Two'},
+          ]}
+        />
+      );
+    });
```

Comment:
This is slick but I don't trust that everyone is comfortable enough with `void` to grok what's happening 😭

Can we stick to the `render()` pattern here so the assertions are easier to understand at a glance?